src/compiler/utils/string_with_sourcemap.ts

```diff
@@ -0,0 +1,165 @@
+import { encode as sourcemap_encode } from "sourcemap-codec";
+
+type MappingSegment =
+	| [number]
+	| [number, number, number, number]
+	| [number, number, number, number, number];
+
+type SourceMappings = {
+	sources: string[];
+	names: string[]; // names are ignored. might change in future
+	mappings: MappingSegment[][];
+};
+
+type SourceLocation = {
+	line: number;
+	column: number;
+};
+
+function last_line_length(s: string) {
+	return s.length - s.lastIndexOf('\n') - 1;
+}
+
+export function sourcemap_add_offset(
+	map: SourceMappings, offset: SourceLocation
+): SourceMappings {
+	return {
+		sources: map.sources.slice(),
+		mappings: map.mappings.map((line, line_idx) =>
+			line.map(seg => {
+				const new_seg = seg.slice() as MappingSegment;
+				if (seg.length >= 4) {
+					new_seg[2] = new_seg[2] + offset.line;
+					if (line_idx == 0)
+						new_seg[3] = new_seg[3] + offset.column;
+				}
+				return new_seg;
+			})
+		)
+	} as SourceMappings;
+}
+
+function merge_tables<T>(this_table: T[], other_table): [T[], number[]] {
+	const new_table = this_table.slice();
+	const idx_map = [];
+	other_table = other_table || [];
+	for (const [other_idx, other_val] of other_table.entries()) {
+		const this_idx = this_table.indexOf(other_val);
+		if (this_idx >= 0) {
+			idx_map[other_idx] = this_idx;
+		} else {
+			const new_idx = new_table.length;
+			new_table[new_idx] = other_val;
+			idx_map[other_idx] = new_idx;
+		}
+	}
+	return [new_table, idx_map];
+}
+
+export class StringWithSourcemap {
+	readonly string: string;
+	readonly map: SourceMappings;
+
+	constructor(string: string, map: SourceMappings) {
+		this.string = string;
+		this.map = map;
+	}
+
+	get_sourcemap() {
+		return {
+			version: 3,
+			sources: this.map.sources.slice(),
+			names: [],
+			mappings: sourcemap_encode(this.map.mappings as any)
+		};
+	}
+
+	concat(other: StringWithSourcemap): StringWithSourcemap {
+		// noop: if one is empty, return the other
+		if (this.string == '') return other;
+		if (other.string == '') return this;
+
+		// combine sources
+		const [sources, new_source_idx] = merge_tables(this.map.sources, other.map.sources);
+		//const [names, new_name_idx] = merge_tables(this.map.names, other.map.names);
```

Comment:
why is this commented out? we should probably either add it back in a working fashion or remove it if unnecessary

Reply:
see line 10:
```ts
 	names: string[]; // names are ignored. might change in future
```
and line 72 - names is always returned empty
```ts
	get_sourcemap() {
		return {
			version: 3,
			sources: this.map.sources.slice(),
			names: [],
			mappings: sourcemap_encode(this.map.mappings as any)
		};
	}
```
so the result is the same, i just made it more clear + disabled dead code

ideally we should support sourcemaps with a `names` field
names is an array of original token/symbol names
that were renamed in the transformation, for example by terser.minify

this could be made an option for svelte.preprocess, like `{ removeSourcemapNames: true }`
so users can avoid post-processing the sourcemap file